{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "2eb957c8-48c6-4043-ad1e-ee16ffa5d2f2",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "This script will take some time to run.\n",
      "Statements will be printed throughout to update on pprogress.\n",
      "Estimated run time:\n",
      "Block 1: 15 hours\n",
      "Block 2: 10 minutes\n",
      "Block 3: 1 minute\n",
      "Data ingested.\n",
      "Beginning Block 1: Generating Kozaks\n",
      "RNA Sequences ready.\n",
      "Protein Sequences ready.\n",
      "reading frames translated\n",
      "Reading frame 1 method 1 done.\n",
      "Reading frame 2 method 1 done.\n",
      "Reading frame 3 method 1 done.\n",
      "Method 1 done.\n",
      "method 2 done.\n",
      "Block 1 output saved as output/block_1_output.csv\n",
      "Beginning Block 2: Isolating Kozak Mutations in Clinvar\n",
      "Block 2 output saved as output/block_2_output.csv\n",
      "Beginning Block 3: Merging the Data and Apply Mutation\n",
      "Block 3 output saved as output/clinvar_kozak_mutations.csv\n"
     ]
    }
   ],
   "source": [
    "# Packages\n",
    "import os\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "\n",
    "# Custom Functions\n",
    "\n",
    "# Amino Acid Translator\n",
    "def amino_acids(Sequence, iinput):\n",
    "    # Please note that this uses the python indexing.  0 is reading frame 1, 1 is reading frame 2, and 2 is reading frame 3 for iinput.\n",
    "    i = int(iinput)\n",
    "    Peptide = []\n",
    "    codon_dict = {'TTT':'F', 'TTC':'F', 'TTA':'L', 'TTG':'L', 'CTT':'L', 'CTC':'L', 'CTA':'L', 'CTG':'L',\n",
    "                  'ATT':'I', 'ATC':'I', 'ATA':'I', 'ATG':'M', 'GTT':'V', 'GTC':'V', 'GTA':'V', 'GTG':'V',\n",
    "                  'TCT':'S', 'TCC':'S', 'TCA':'S', 'TCG':'S', 'CCT':'P', 'CCC':'P', 'CCA':'P', 'CCG':'P',\n",
    "                  'ACT':'T', 'ACC':'T', 'ACA':'T', 'ACG':'T', 'GCT':'A', 'GCC':'A', 'GCA':'A', 'GCG':'A',\n",
    "                  'TAT':'Y', 'TAC':'Y', 'TAA':'*', 'TAG':'*', 'CAT':'H', 'CAC':'H', 'CAA':'Q', 'CAG':'Q',\n",
    "                  'AAT':'N', 'AAC':'N', 'AAA':'K', 'AAG':'K', 'GAT':'D', 'GAC':'D', 'GAA':'E', 'GAG':'E',\n",
    "                  'TGT':'C', 'TGC':'C', 'TGA':'*', 'TGG':'W', 'CGT':'R', 'CGC':'R', 'CGA':'R', 'CGG':'R',\n",
    "                  'AGT':'S', 'AGC':'S', 'AGA':'R', 'AGG':'R', 'GGT':'G', 'GGC':'G', 'GGA':'G', 'GGG':'G',\n",
    "                  # entries for N nucleotides\n",
    "                  'TTN':'X', 'CTN':'L', 'ATN':'X', 'GTN':'V', 'TCN':'S', 'CCN':'P', 'ACN':'T', 'GCN':'A',\n",
    "                  'TAN':'X', 'CAN':'X', 'AAN':'X', 'GAN':'X', 'TGN':'X', 'CGN':'R', 'AGN':'X', 'GGN':'G',\n",
    "                  'TNT':'X', 'TNC':'X', 'TNA':'X', 'TNG':'X', 'CNT':'X', 'CNC':'X', 'CNA':'X', 'CNG':'X',\n",
    "                  'ANT':'X', 'ANC':'X', 'ANA':'X', 'ANG':'X', 'GNT':'X', 'GNC':'X', 'GNA':'X', 'GNG':'X',\n",
    "                  'NTT':'X', 'NTC':'X', 'NTA':'X', 'NTG':'X', 'NCT':'X', 'NCC':'X', 'NCA':'X', 'NCG':'X',\n",
    "                  'NAT':'X', 'NAC':'X', 'NAA':'X', 'NAG':'X', 'NGT':'X', 'NGC':'X', 'NGA':'X', 'NGG':'X',\n",
    "                  'NNT':'X', 'NNA':'X', 'NNC':'X', 'NNG':'X', 'NTN':'X', 'NAN':'X', 'NCN':'X', 'NGN':'X',\n",
    "                  'TNN':'X', 'ANN':'X', 'CNN':'X', 'GNN':'X', 'NNN':'X',\n",
    "                  # entries for Y nucleotides\n",
    "                  'TTY':'F', 'CTY':'L', 'ATY':'I', 'GTY':'V', 'TCY':'S', 'CCY':'P', 'ACY':'T', 'GCY':'A',\n",
    "                  'TAY':'Y', 'CAY':'H', 'AAY':'N', 'GAY':'D', 'TGY':'C', 'CGY':'R', 'AGY':'S', 'GGY':'G',\n",
    "                  'TYT':'X', 'TYC':'X', 'TYA':'X', 'TYG':'X', 'CYT':'X', 'CYC':'X', 'CYA':'X', 'CYG':'X',\n",
    "                  'AYT':'X', 'AYC':'X', 'AYA':'X', 'AYG':'X', 'GYT':'X', 'GYC':'X', 'GYA':'X', 'GYG':'X',\n",
    "                  'YTT':'X', 'YTC':'X', 'YTA':'L', 'YTG':'L', 'YCT':'X', 'YCC':'X', 'YCA':'X', 'YCG':'X',\n",
    "                  'YAT':'X', 'YAC':'X', 'YAA':'X', 'YAG':'X', 'YGT':'X', 'YGC':'X', 'YGA':'X', 'YGG':'X',\n",
    "                  'YYT':'X', 'YYA':'X', 'YYC':'X', 'YYG':'X', 'YTY':'X', 'YAY':'X', 'YCY':'X', 'YGY':'X',\n",
    "                  'TYY':'X', 'AYY':'X', 'CYY':'X', 'GYY':'X', 'YYY':'X',\n",
    "                  # entries R nucleotides\n",
    "                  'TTR':'L', 'CTR':'L', 'ATR':'X', 'GTR':'V', 'TCR':'S', 'CCR':'P', 'ACR':'T', 'GCR':'A',\n",
    "                  'TAR':'*', 'CAR':'Q', 'AAR':'K', 'GAR':'E', 'TGR':'X', 'CGR':'R', 'AGR':'R', 'GGR':'G',\n",
    "                  'TRT':'X', 'TRC':'X', 'TRA':'*', 'TRG':'X', 'CRT':'X', 'CRC':'X', 'CRA':'X', 'CRG':'X',\n",
    "                  'ART':'X', 'ARC':'X', 'ARA':'X', 'ARG':'X', 'GRT':'X', 'GRC':'X', 'GRA':'X', 'GRG':'X',\n",
    "                  'RTT':'X', 'RTC':'X', 'RTA':'L', 'RTG':'L', 'RCT':'X', 'RCC':'X', 'RCA':'X', 'RCG':'X',\n",
    "                  'RAT':'X', 'RAC':'X', 'RAA':'X', 'RAG':'X', 'RGT':'X', 'RGC':'X', 'RGA':'X', 'RGG':'X',\n",
    "                  'RRT':'X', 'RRA':'X', 'RRC':'X', 'RRG':'X', 'RTR':'X', 'RAR':'X', 'RCR':'X', 'RGR':'X',\n",
    "                  'TRR':'X', 'ARR':'X', 'CRR':'X', 'GRR':'X', 'RRR':'X',\n",
    "                  # entries V nucleotides\n",
    "                  'TTV':'X', 'CTV':'L', 'ATV':'X', 'GTV':'V', 'TCV':'S', 'CCV':'P', 'ACV':'T', 'GCV':'A',\n",
    "                  'TAV':'X', 'CAV':'X', 'AAV':'X', 'GAV':'X', 'TGV':'X', 'CGV':'R', 'AGV':'X', 'GGV':'G',\n",
    "                  'TVT':'X', 'TVC':'X', 'TVA':'X', 'TVG':'X', 'CVT':'X', 'CVC':'X', 'CVA':'X', 'CVG':'X',\n",
    "                  'AVT':'X', 'AVC':'X', 'AVA':'X', 'AVG':'X', 'GVT':'X', 'GVC':'X', 'GVA':'X', 'GVG':'X',\n",
    "                  'VTT':'X', 'VTC':'X', 'VTA':'X', 'VTG':'X', 'VCT':'X', 'VCC':'X', 'VCA':'X', 'VCG':'X',\n",
    "                  'VAT':'X', 'VAC':'X', 'VAA':'X', 'VAG':'X', 'VGT':'X', 'VGC':'X', 'VGA':'X', 'VGG':'X',\n",
    "                  'VVT':'X', 'VVA':'X', 'VVC':'X', 'VVG':'X', 'VTV':'X', 'VAV':'X', 'VCV':'X', 'VGV':'X',\n",
    "                  'TVV':'X', 'AVV':'X', 'CVV':'X', 'GVV':'X', 'VVV':'X',\n",
    "                  # entries B nucleotides\n",
    "                  'TTB':'X', 'CTB':'L', 'ATB':'X', 'GTB':'V', 'TCB':'S', 'CCB':'P', 'ACB':'T', 'GCB':'A',\n",
    "                  'TAB':'X', 'CAB':'X', 'AAB':'X', 'GAB':'X', 'TGB':'X', 'CGB':'R', 'AGB':'X', 'GGB':'G',\n",
    "                  'TBT':'X', 'TBC':'X', 'TBA':'X', 'TBG':'X', 'CBT':'X', 'CBC':'X', 'CBA':'X', 'CBG':'X',\n",
    "                  'ABT':'X', 'ABC':'X', 'ABA':'X', 'ABG':'X', 'GBT':'X', 'GBC':'X', 'GBA':'X', 'GBG':'X',\n",
    "                  'BTT':'X', 'BTC':'X', 'BTA':'X', 'BTG':'X', 'BCT':'X', 'BCC':'X', 'BCA':'X', 'BCG':'X',\n",
    "                  'BAT':'X', 'BAC':'X', 'BAA':'X', 'BAG':'X', 'BGT':'X', 'BGC':'X', 'BGA':'X', 'BGG':'X',\n",
    "                  'BBT':'X', 'BBA':'X', 'BBC':'X', 'BBG':'X', 'BTB':'X', 'BAB':'X', 'BCB':'X', 'BGB':'X',\n",
    "                  'TBB':'X', 'ABB':'X', 'CBB':'X', 'GBB':'X', 'BBB':'X'}\n",
    "    while i < len(Sequence):\n",
    "        if len(Sequence[i:i+3]) == 3:\n",
    "            Peptide.append(codon_dict[Sequence[i:i+3]])\n",
    "        i = i+3\n",
    "    return(''.join(Peptide))\n",
    "\n",
    "# Find Index of max length seq\n",
    "def Find_Index(List, MaxLength):\n",
    "    for index, seq in enumerate(List):\n",
    "        if len(seq) == MaxLength:\n",
    "            return index\n",
    "\n",
    "# read text file\n",
    "def readmyfile(filename):\n",
    "    with open(filename, 'r') as MyData:\n",
    "        lines = []\n",
    "        for fileline in MyData:\n",
    "            lines.append(fileline.rstrip())\n",
    "    return(lines)\n",
    "\n",
    "# find kozak through exact string match between RNA translation and protein\n",
    "def find_kozak_method_1(translation,readframe,rnaseq):\n",
    "    rf = readframe-1\n",
    "    possible_kozak = []\n",
    "    for prot in prot_sequences:\n",
    "        if translation.find(prot) != -1:\n",
    "            possible_kozak.append(rnaseq[translation.find(prot)*3+rf-6:translation.find(prot)*3+rf])\n",
    "            break #this will cause us to only call the first match.  Some may have multiple matches.\n",
    "                  #however, since this would mean 25,262,185,494 checks per reading frame at time of\n",
    "                  #writing, I am reducing to limit computational load.  However, note that removing\n",
    "                  #will break downstream code.\n",
    "    return(possible_kozak)\n",
    "\n",
    "# find kozak through through 'kozak' for longest string in any rf\n",
    "def find_kozak_method_2(translation1, # reading frame 1\n",
    "                        translation2, # reading frame 2\n",
    "                        translation3, # reading frame 3\n",
    "                        rnaseq):\n",
    "    # set up intermediate lists\n",
    "    maxseqlens = []\n",
    "    # set up rf1\n",
    "    seqs1 = translation1.split('*')\n",
    "    seqlens1 = []   \n",
    "    for seq in seqs1:\n",
    "        seqlens1.append(len(seq))\n",
    "    maxseqlens.append(max(seqlens1))\n",
    "    # set up rf2\n",
    "    seqs2 = translation2.split('*')\n",
    "    seqlens2 = []\n",
    "    for seq in seqs2:\n",
    "        seqlens2.append(len(seq))\n",
    "    maxseqlens.append(max(seqlens2))\n",
    "    # set up rf3\n",
    "    seqs3 = translation3.split('*')\n",
    "    seqlens3 = []\n",
    "    for seq in seqs3:\n",
    "        seqlens3.append(len(seq))\n",
    "    maxseqlens.append(max(seqlens3))\n",
    "    \n",
    "    # case of rf1 has max\n",
    "    if maxseqlens.index(max(maxseqlens)) == 0:\n",
    "        longest_string = seqs1[seqlens1.index(max(seqlens1))]\n",
    "        if longest_string.find('M') == -1:\n",
    "            return(np.nan) # no methionine, aka no translation\n",
    "        else:\n",
    "            longest_string_translated = longest_string[longest_string.find('M'):]\n",
    "            return(rnaseq[translation1.find(longest_string_translated)*3+0-6:translation1.find(longest_string_translated)*3+0])\n",
    "    \n",
    "    # case of rf2 has max\n",
    "    elif maxseqlens.index(max(maxseqlens)) == 1:\n",
    "        longest_string = seqs2[seqlens2.index(max(seqlens2))]\n",
    "        if longest_string.find('M') == -1:\n",
    "            return(np.nan) # no methionine, aka no translation\n",
    "        else:\n",
    "            longest_string_translated = longest_string[longest_string.find('M'):]\n",
    "            return(rnaseq[translation2.find(longest_string_translated)*3+1-6:translation2.find(longest_string_translated)*3+1])\n",
    "    \n",
    "    # case of rf3 has max\n",
    "    elif maxseqlens.index(max(maxseqlens)) == 2:\n",
    "        longest_string = seqs3[seqlens3.index(max(seqlens3))]\n",
    "        if longest_string.find('M') == -1:\n",
    "            return(np.nan) # no methionine, aka no translation\n",
    "        else:\n",
    "            longest_string_translated = longest_string[longest_string.find('M'):]\n",
    "            return(rnaseq[translation3.find(longest_string_translated)*3+2-6:translation3.find(longest_string_translated)*3+2])\n",
    "    \n",
    "    # return NAN if something didn't work otherwise\n",
    "    else:\n",
    "        return(np.nan)\n",
    "    \n",
    "# merge outputs for each reading frame of finding kozak through method 1\n",
    "def unify_method_1(rf1,rf2,rf3):\n",
    "    if str(rf1) != '[]':\n",
    "        return(str(rf1).replace('[','').replace(']','').replace(\"'\",''))\n",
    "    elif str(rf2) != '[]':\n",
    "        return(str(rf2).replace('[','').replace(']','').replace(\"'\",''))\n",
    "    elif str(rf3) != '[]':\n",
    "        return(str(rf3).replace('[','').replace(']','').replace(\"'\",''))\n",
    "    else:\n",
    "        return(np.nan)\n",
    "\n",
    "# extract NM identifier in block 1\n",
    "def extract_nm(input):\n",
    "    return(input[input.find('>')+1:input.find(' ')])\n",
    "\n",
    "# extract NM identifier in block 2\n",
    "def extract_nm2(input):\n",
    "    return(input[:input.find('(')])\n",
    "\n",
    "# outputs location of mutations around kozak sequence\n",
    "def find_location(mut):\n",
    "    if mut[2] == '+':\n",
    "        return('intron')\n",
    "    elif mut[2] ==  '-':\n",
    "        return('intron')\n",
    "    else:\n",
    "        return('kozak')\n",
    "    \n",
    "# outputs old kozak sequence and updated mutated kozak sequence.  Supplies appropriate error or \n",
    "# 'errors' if excluded situations or failures are encountered.\n",
    "def mutate_kozak(m1_kozak, m2_kozak, mut_location, mut_type, mut_details):\n",
    "    # return error:no_kozak_found if no kozak from method 1 or method 2.\n",
    "    if pd.isnull(m1_kozak) == True:\n",
    "        if pd.isnull(m2_kozak) == True:\n",
    "            return('error:no_kozak_found','error:no_kozak_found')\n",
    "    # return error:intron_mutation if mutation in intron.  We will not be\n",
    "    # handling such cases due to complexity and scope.\n",
    "    if mut_location == 'intron':\n",
    "        return('error:intron_mutation','error:intron_mutation')\n",
    "    # return error:microsatellite if mutation is a microsatellite.  We will\n",
    "    # not be handling such cases due to complexity and scope.\n",
    "    if mut_type == 'Microsatellite':\n",
    "        return('error:microsatellite','error:microsatellite')\n",
    "    # making variables to simplify reading of rest of this function.\n",
    "    position = int(mut_details[0:2]) # make sure this is a number!\n",
    "    oldnt = mut_details[2]\n",
    "    # handle case of duplication.  There is only one at type of writing, but\n",
    "    # will be generalizable in case more found in future.  In principle, would\n",
    "    # be same for a deletion, but would require going back into the RNA to collect.\n",
    "    if mut_type == 'Duplication':\n",
    "        if pd.isnull(m1_kozak) == False:\n",
    "            if m1_kozak[position] == oldnt:\n",
    "                return(m1_kozak,m1_kozak[-5:position] + 2*newnt + m1_kozak[position+1:])\n",
    "        elif pd.isnull(m2_kozak) == False:\n",
    "            if m2_kozak[position] == oldnt:\n",
    "                return(m2_kozak,m2_kozak[-5:position] + 2*newnt + m2_kozak[position+1:])\n",
    "        else:\n",
    "            # return error:mutation_not_match_kozak when mutation doesn't match either method 1 or 2 kozak.\n",
    "            return('error:mutation_not_match_kozak','error:mutation_not_match_kozak')\n",
    "    # making variables to simplify reading of rest of this function.      \n",
    "    newnt = mut_details[4]\n",
    "    if mut_type == 'single nucleotide variant':\n",
    "        if pd.isnull(m1_kozak) == False:\n",
    "            if m1_kozak[position] == oldnt:\n",
    "                if position == -1:\n",
    "                    return(m1_kozak, m1_kozak[-6:position] + newnt)\n",
    "                else:\n",
    "                    return(m1_kozak, m1_kozak[-6:position] + newnt + m1_kozak[position+1:])\n",
    "            elif pd.isnull(m2_kozak) == False:\n",
    "                if m2_kozak[position] == oldnt:\n",
    "                    if position == -1:\n",
    "                        return(m2_kozak, m2_kozak[-6:position] + newnt)\n",
    "                    else:\n",
    "                        return(m2_kozak, m2_kozak[-6:position] + newnt + m2_kozak[position+1:])\n",
    "                else:\n",
    "                    return('error:mutation_not_match_kozak','error:mutation_not_match_kozak')\n",
    "            else:\n",
    "                return('error:mutation_not_match_kozak','error:mutation_not_match_kozak')\n",
    "        elif pd.isnull(m2_kozak) == False:\n",
    "            if m2_kozak[position] == oldnt:\n",
    "                if position == -1:\n",
    "                    return(m2_kozak, m2_kozak[-6:position] + newnt)\n",
    "                else:\n",
    "                    return(m2_kozak, m2_kozak[-6:position] + newnt + m2_kozak[position+1:])\n",
    "            else:\n",
    "                return('error:mutation_not_match_kozak','error:mutation_not_match_kozak')\n",
    "        else:\n",
    "            # return error:mutation_not_match_kozak when mutation doesn't match either method 1 or 2 kozak.\n",
    "            return('error:mutation_not_match_kozak','error:mutation_not_match_kozak')\n",
    "    else:\n",
    "        # output if everything else failed\n",
    "        return('error:function_failed','error:function_failed')\n",
    "    \n",
    "### Block 0 ###\n",
    "\n",
    "# Below parses the RNA and Protein file from\n",
    "# https://www.ncbi.nlm.nih.gov/genome/guide/human/ for GRCh38, and prepares\n",
    "# a background list (info about the sequence data) and a sequence file (the actual sequence data).\n",
    "# Use the Refseq Transcripts and Refseq Proteins downloads in this situation.\n",
    "# --> https://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh38_latest/refseq_identifiers/GRCh38_latest_rna.fna.gz\n",
    "# --> https://ftp.ncbi.nlm.nih.gov/refseq/H_sapiens/annotation/GRCh38_latest/refseq_identifiers/GRCh38_latest_protein.faa.gz\n",
    "# For later, we will also need the variant summary text file from Clinvar.  This can be found at:\n",
    "# https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/archive/\n",
    "# select the variant_summary_(date).txt.gz with the most current date for this.\n",
    "# On mac, after downloading, go to terminal and use: gunzip (drag file after space into terminal) to unzip each of these files.\n",
    "# On windows, go to terminal and use: tar -xvzf (filepath) to unzip each of these files\n",
    "# make sure files are in a folder titled 'data' at same directory position as this script\n",
    "\n",
    "print('This script will take some time to run.')\n",
    "print('Statements will be printed throughout to update on progress.')\n",
    "print('Estimated run time:')\n",
    "print('     * Block 1: 15 hours')\n",
    "print('     * Block 2: 10 minutes')\n",
    "print('     * Block 3: 1 minute')\n",
    "\n",
    "GRCH38_rna = readmyfile('data/GRCh38_latest_rna.fna')\n",
    "GRCH38_prot = readmyfile('data/GRCh38_latest_protein.faa')\n",
    "clinvar_dat = pd.read_csv('data/variant_summary_2025-02.txt', sep='\\t', header = 0, low_memory = False)\n",
    "print('')\n",
    "print('Data ingested.')\n",
    "print('')\n",
    "\n",
    "# makes output folder\n",
    "if not os.path.exists('output'):\n",
    "    os.makedirs('output')\n",
    "\n",
    "### Block 1 ###\n",
    "\n",
    "## Code Block 1 is designed to generate KOZAK data by utilizing the data from GRCh38.\n",
    "\n",
    "print('Beginning Block 1: Generating Kozaks')\n",
    "i = 0\n",
    "rna_headers = []\n",
    "rna_sequences = []\n",
    "temp_sequence = []\n",
    "while i < len(GRCH38_rna):\n",
    "    if GRCH38_rna[i][0] == '>':\n",
    "        rna_sequences.append(''.join(temp_sequence))\n",
    "        temp_sequence = []\n",
    "        rna_headers.append(GRCH38_rna[i])\n",
    "    else:\n",
    "        temp_sequence.append(GRCH38_rna[i])\n",
    "    i +=1\n",
    "rna_sequences.append(''.join(temp_sequence))\n",
    "rna_sequences = rna_sequences[1:] # remove the empty first row that was generated in seqs\n",
    "print('RNA Sequences ready.')\n",
    "\n",
    "i = 0\n",
    "prot_headers = []\n",
    "prot_sequences = []\n",
    "temp_sequence = []\n",
    "while i < len(GRCH38_prot):\n",
    "    if GRCH38_prot[i][0] == '>':\n",
    "        prot_sequences.append(''.join(temp_sequence))\n",
    "        temp_sequence = []\n",
    "        prot_headers.append(GRCH38_prot[i])\n",
    "    else:\n",
    "        temp_sequence.append(GRCH38_prot[i])\n",
    "    i +=1\n",
    "prot_sequences.append(''.join(temp_sequence))\n",
    "prot_sequences = prot_sequences[1:] # remove the empty first row that was generated in seqs\n",
    "print('Protein Sequences ready.')\n",
    "\n",
    "# Garbage Collection\n",
    "del(temp_sequence)\n",
    "del(GRCH38_rna)\n",
    "del(GRCH38_prot)\n",
    "\n",
    "kozak_df = pd.DataFrame({\n",
    "    'grch38_headers': rna_headers,\n",
    "    'grch38_rna_seqs': rna_sequences})\n",
    "\n",
    "        ## Section 2: Find Kozak Sequences ##\n",
    "\n",
    "## Approach 1\n",
    "# Here, we will attempt to find the KOZAK sequence by translating the RNA sequences and\n",
    "# cross referencing them with the protein dataset.\n",
    "\n",
    "# create translations in each reading frame\n",
    "kozak_df['reading_frame_1'] = kozak_df.apply(lambda x: amino_acids(x.grch38_rna_seqs,0), axis = 1)\n",
    "kozak_df['reading_frame_2'] = kozak_df.apply(lambda x: amino_acids(x.grch38_rna_seqs,1), axis = 1)\n",
    "kozak_df['reading_frame_3'] = kozak_df.apply(lambda x: amino_acids(x.grch38_rna_seqs,2), axis = 1)\n",
    "print('reading frames translated')\n",
    "\n",
    "# find string matches\n",
    "kozak_df['rf1_m1_kozaks'] = kozak_df.apply(lambda x: find_kozak_method_1(x.reading_frame_1,\n",
    "                                                                         1,\n",
    "                                                                         x.grch38_rna_seqs),axis = 1)\n",
    "print('Reading frame 1 method 1 done.')\n",
    "kozak_df['rf2_m1_kozaks'] = kozak_df.apply(lambda x: find_kozak_method_1(x.reading_frame_2,\n",
    "                                                                         2,\n",
    "                                                                         x.grch38_rna_seqs),axis = 1)\n",
    "print('Reading frame 2 method 1 done.')\n",
    "kozak_df['rf3_m1_kozaks'] = kozak_df.apply(lambda x: find_kozak_method_1(x.reading_frame_3,\n",
    "                                                                         3,\n",
    "                                                                         x.grch38_rna_seqs), axis = 1)\n",
    "print('Reading frame 3 method 1 done.')\n",
    "\n",
    "kozak_df['m1_kozak'] = kozak_df.apply(lambda x: unify_method_1(x.rf1_m1_kozaks,\n",
    "                                                               x.rf2_m1_kozaks,\n",
    "                                                               x.rf3_m1_kozaks),axis = 1)\n",
    "print('Method 1 done.')\n",
    "\n",
    "## Approach 2\n",
    "# This approach searches for the longest protein sequence in each reading frame, and finds\n",
    "# the kozak for that sequence.  This is less optimal as an approach, but it's included to be comprehensive.\n",
    "\n",
    "kozak_df['m2_kozak'] = kozak_df.apply(lambda x: find_kozak_method_2(x.reading_frame_1,\n",
    "                                                                    x.reading_frame_2,\n",
    "                                                                    x.reading_frame_3,\n",
    "                                                                    x.grch38_rna_seqs), axis = 1)\n",
    "print('method 2 done.')\n",
    "\n",
    "# pull out nm number, which will be used later to merge with clinvar data\n",
    "kozak_df['nm'] = kozak_df.apply(lambda x: extract_nm(x.grch38_headers), axis = 1)\n",
    "\n",
    "# replace field that's entirely space (or empty) with NaN\n",
    "kozak_df = kozak_df.replace('', np.nan, regex = True)\n",
    "\n",
    "kozak_df.to_csv('output/block_1_output.csv')\n",
    "print('Block 1 output saved as output/block_1_output.csv')\n",
    "\n",
    "# Garbage Collection\n",
    "del(rna_headers)\n",
    "del(rna_sequences)\n",
    "del(prot_headers)\n",
    "del(prot_sequences)\n",
    "del(i)\n",
    "\n",
    "### Block 2 ###\n",
    "\n",
    "## Code Block 2 is designed to set up the dataframe of clinvar variants we will look at the kozak mutations of.\n",
    "\n",
    "print('Beginning Block 2: Isolating Kozak Mutations in Clinvar')\n",
    "\n",
    "# To start, we will need to remove all GRCh37 entries from the data, which are found in the Assembly column.\n",
    "clinvar_dat = clinvar_dat[clinvar_dat['Assembly'] == 'GRCh38']\n",
    "\n",
    "# Now, we will remove all columns that are extraneous for our purposes.\n",
    "clinvar_dat = clinvar_dat[['Type','Name','GeneID','GeneSymbol',\n",
    "                           'HGNC_ID','ClinicalSignificance','Chromosome',\n",
    "                           'Start','Stop','Oncogenicity']]\n",
    "\n",
    "# The two main columns that we will want to start with are Type and Name.\n",
    "# Type contains mutation type (indel/SNV/etc.)\n",
    "# Name contains the actual mutation information that we will want to get at.\n",
    "# e.g. NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)\n",
    "# :c. comes right before the mutation\n",
    "# next is the position(s) that are affected (3136)\n",
    "# then the change(s) (G --> A)\n",
    "# what we will do first is make a new column off of the name column with the mutation data after :c.,\n",
    "# after which we will isolate out mutations affecting the kozak positions.\n",
    "clinvar_dat['mutation'] = clinvar_dat.apply(lambda x: x.Name[x.Name.find(':c.')+3:], axis = 1)\n",
    "clinvar_dat = clinvar_dat[clinvar_dat['mutation'].str.startswith(('-1A','-1T','-1C','-1G','-2A','-2T','-2C','-2G',\n",
    "                                                                  '-3A','-3T','-3C','-3G','-4A','-4T','-4C','-4G',\n",
    "                                                                  '-5A','-5T','-5C','-5G','-6A','-6T','-6C','-6G',\n",
    "                                                                  '-1-','-1+','-2-','-2+','-3-','-3+','-4-','-4+',\n",
    "                                                                  '-5-','-5+','-6+'))]\n",
    "# To simplify later steps, we will identify which mutations affect introns here.\n",
    "clinvar_dat['mutation_location'] = clinvar_dat.apply(lambda x: find_location(x.mutation), axis = 1)\n",
    "\n",
    "clinvar_dat['nm'] = clinvar_dat.apply(lambda x: extract_nm2(x.Name), axis = 1)\n",
    "\n",
    "# replace field that's entirely space (or empty) with NaN\n",
    "clinvar_dat = clinvar_dat.replace('', np.nan, regex = True)\n",
    "\n",
    "clinvar_dat.to_csv('output/block_2_output.csv')\n",
    "print('Block 2 output saved as output/block_2_output.csv')\n",
    "\n",
    "### Block 3 ###\n",
    "\n",
    "## Code Block 3 will merge the two prior dataframes, and apply kozak mutations.\n",
    "\n",
    "print('Beginning Block 3: Merging the Data and Apply Mutation')\n",
    "\n",
    "# To merge the data, we will use the NM values.\n",
    "# We will be focusing on using the keys in the clinvar_dat frame.\n",
    "# This will provide a smaller dataframe to work within.\n",
    "final_dat = pd.merge(kozak_df, clinvar_dat, how = 'right', on =  ['nm'])\n",
    "\n",
    "# replace field that's entirely space (or empty) with NaN\n",
    "# especially important here - empty cells cause failure\n",
    "final_dat = final_dat.replace('', np.nan, regex = True)\n",
    "\n",
    "final_dat[['old_kozak','new_kozak']] = final_dat.apply(lambda x: mutate_kozak(x.m1_kozak,\n",
    "                                                                              x.m2_kozak,\n",
    "                                                                              x.mutation_location,\n",
    "                                                                              x.Type,\n",
    "                                                                              x.mutation),\n",
    "                                                       axis = 'columns',\n",
    "                                                       result_type = 'expand')\n",
    "\n",
    "# Save output\n",
    "final_dat.to_csv('output/clinvar_kozak_mutations.csv')\n",
    "print('Block 3 output saved as output/clinvar_kozak_mutations.csv')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
